File: Tries/208_ImplementPrefixTree.cpp
#include <string>

using namespace std;

/**
 * 208. Implement Trie (Prefix Tree)
 * 
 * Implementing a prefix tree in C++
 * This was a good way to learn how to 
 * do memory management in C++ so 
 * woohoo
 */

struct TrieNode {
    TrieNode* children[26];
    bool end = false;

    TrieNode() {
        for (int i = 0; i < 26; i++) {
            children[i] = nullptr;
        }
    }
};

class Trie {
private:
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char i : word) {
            int index = i - 'a';

            if (!node->children[index]) {
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->end = true;
    }
    
    bool search(string word) {
        TrieNode* node = root;

        for (char i : word) {
            int index = i - 'a';

            if (!node->children[index]) {
                return false;
            }

            node = node->children[index];
        }
        if (!node->end) {
            return false;
        }
        return true;
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;

        for (char i : prefix) {
            int index = i - 'a';

            if (!node->children[index]) {
                return false;
            }

            node = node->children[index];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */


/**
 * The next try for the trie (heh) invloved this
 * it was pretty much the exact same lmaoo
 * but hey slightly more memory efficient
 */

class TrieNode {
public:
    TrieNode* children[26];
    bool end = false;

    TrieNode() {
        for (int i = 0; i < 26; ++i) {
            children[i] = nullptr;
        }
    }
};
    
class Trie {
public:
    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* curr = root;
        for (char& c : word) {
            int index = c - 'a';
            if (curr->children[index] == nullptr) {
                curr->children[index] = new TrieNode();
            }
            curr = curr->children[index];
        }
        curr->end = true;
        return;
    }
    
    bool search(string word) {
        TrieNode* curr = root;
        for (char& c : word) {
            int index = c - 'a';
            curr = curr->children[c - 97];
            if (curr == nullptr) { return false; }
        }
        return curr->end;
    }
    
    bool startsWith(string prefix) {
        TrieNode* curr = root;
        for (char& c : prefix) {
            curr = curr->children[c - 97];
            if (curr == nullptr) { return false; }
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: Tries/208_ImplementPrefixTree.py
class TrieNode:
    """
    A Trie Node to keep track of each individual letter and what
    follows it
    """
    def __init__(self):
        self.dic = {}
        self.end = False

class Trie:
    # initialize the root node which is nothing really it just holds the first letters
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserting a word into the Trie
        It will still fail if you try to add a word that
        has already been added, it just won't do anything
        meaningful
        
        param word: the word we are adding to the tree
        """
        currNode = self.root
        for char in word:
            if char not in currNode.dic:
                currNode.dic[char] = TrieNode()
            currNode = currNode.dic[char]
        currNode.end = True

    def search(self, word: str) -> bool:
        """
        Searching for a word in the Trie 
        If there is a word but it was not specifically
        added it will return false since that letter
        was not considered the end of the word

        param word: the word we are searching for
        """
        currNode = self.root
        for char in word:
            if char not in currNode.dic:
                return False
            currNode = currNode.dic[char]
        if currNode.end != True:
            return False
        return True
        

    def startsWith(self, prefix: str) -> bool:
        """
        Searching for the start of the word
        If the letters are an entire word or the start
        of the word then it will return true, otherwise
        false

        param prefix: the word we are searching for
        """
        currNode = self.root
        for char in prefix:
            if char not in currNode.dic:
                return False
            currNode = currNode.dic[char]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)